src/Microsoft.AspNetCore.Mvc.Core/Internal/DefaultApplicationModelProvider.cs

```diff
@@ -213,24 +217,20 @@ protected virtual PropertyModel CreatePropertyModel(PropertyInfo propertyInfo)
                 throw new ArgumentNullException(nameof(propertyInfo));
             }
 
-            // CoreCLR returns IEnumerable<Attribute> from GetCustomAttributes - the OfType<object>
-            // is needed to so that the result of ToArray() is object
             var attributes = propertyInfo.GetCustomAttributes(inherit: true);
-            var propertyModel = new PropertyModel(propertyInfo, attributes);
-            var bindingInfo = BindingInfo.GetBindingInfo(attributes);
-            if (bindingInfo != null)
+            var propertyModel = new PropertyModel(propertyInfo, attributes)
             {
-                propertyModel.BindingInfo = bindingInfo;
-            }
-            else if (IsFormFileType(propertyInfo.PropertyType))
+                PropertyName = propertyInfo.Name
+            };
+
+            var modelMetadata = _modelMetadataProvider.GetMetadataForProperty(propertyInfo.DeclaringType, propertyInfo.Name);
+            var bindingInfo = BindingInfo.GetBindingInfo(attributes, modelMetadata);
+            if (bindingInfo == null && IsFormFileType(propertyInfo.PropertyType))
             {
-                propertyModel.BindingInfo = new BindingInfo
-                {
-                    BindingSource = BindingSource.FormFile,
-                };
+                bindingInfo = new BindingInfo { BindingSource = BindingSource.FormFile, };
             }
 
-            propertyModel.PropertyName = propertyInfo.Name;
+            propertyModel.BindingInfo = bindingInfo;
```

Comment:
I think this is a good change overall, but we should dedupe this with the rest of the code that doesn't treat bindinginfo as authoritative.